cmd/embedded-cluster/join.go

```diff
@@ -368,3 +393,100 @@ func runK0sInstallCommand(fullcmd string) error {
+	if err != nil {
+		return fmt.Errorf("unable to create kube client: %w", err)
+	}
+	embeddedclusterv1beta1.AddToScheme(kcli.Scheme())
+	isHA, err := isHAEnabled(ctx, kcli)
+	if err != nil {
+		return fmt.Errorf("unable to check if HA is enabled: %w", err)
+	}
+	if isHA {
+		return nil
+	}
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("unable to get hostname: %w", err)
+	}
+	numNodes, err := waitForNode(ctx, kcli, hostname)
+	if err != nil {
+		return fmt.Errorf("unable to wait for node: %w", err)
+	}
+	if numNodes < 3 {
+		return nil
+	}
+	shouldEnableHA := prompts.New().Confirm("Do you want to enable high availability?", false)
+	if !shouldEnableHA {
+		return nil
+	}
+	return enableHA(ctx, kcli)
+}
+
+// isHAEnabled checks if high availability is already enabled in the cluster.
+func isHAEnabled(ctx context.Context, kcli client.Client) (bool, error) {
+	installation, err := kubeutils.GetLatestInstallation(ctx, kcli)
+	if err != nil {
+		return false, fmt.Errorf("unable to get latest installation: %w", err)
+	}
+	return installation.Spec.HighAvailability, nil
+}
+
+// waitForNode waits for the node to be registered in the k8s cluster.
+// It returns the total number of nodes in the cluster.
+func waitForNode(ctx context.Context, kcli client.Client, name string) (int, error) {
+	loading := spinner.Start()
+	defer loading.Close()
+	loading.Infof("Waiting for %s node to join the cluster", defaults.BinaryName())
+	backoff := wait.Backoff{Steps: 60, Duration: 5 * time.Second, Factor: 1.0, Jitter: 0.1}
+	var nodes corev1.NodeList
+	var lasterr error
+	if err := wait.ExponentialBackoffWithContext(
+		ctx, backoff, func(ctx context.Context) (bool, error) {
+			if err := kcli.List(ctx, &nodes); err != nil {
+				lasterr = fmt.Errorf("unable to list nodes: %v", err)
+				return false, nil
+			}
+			for _, node := range nodes.Items {
+				if node.Name == name {
+					return true, nil
+				}
+			}
+			lasterr = fmt.Errorf("node %s not found", name)
+			return false, nil
+		},
+	); err != nil {
+		if lasterr != nil {
+			return 0, fmt.Errorf("timed out waiting for node %s: %w", name, lasterr)
+		} else {
+			return 0, fmt.Errorf("timed out waiting for node %s", name)
+		}
+	}
+	loading.Infof("Node has joined the cluster!")
+	return len(nodes.Items), nil
+}
+
+// enableHA enables high availability in the installation object
+// and waits for the migration to be complete.
+func enableHA(ctx context.Context, kcli client.Client) error {
+	loading := spinner.Start()
+	defer loading.Close()
+	loading.Infof("Enabling high availability")
+	installation, err := kubeutils.GetLatestInstallation(ctx, kcli)
+	if err != nil {
+		return fmt.Errorf("unable to get latest installation: %w", err)
+	}
+	installation.Spec.HighAvailability = true
+	if err := kcli.Update(ctx, installation); err != nil {
+		return fmt.Errorf("unable to update installation: %w", err)
+	}
+
+	// TODO: how should we wait for HA migration to complete?
```

Comment:
Maybe look for the condition on the installation object? Looking at https://github.com/replicatedhq/embedded-cluster-operator/pull/196

Reply:
Yes